banner_ics.php

```diff
@@ -63,20 +64,47 @@ public function process_attachment(&$content, &$message, &$a)
             // Make sure we have events
             if (!$ical->hasEvents()) return;
 
+            // Crete timezone objects for calendar timezone and RoundCube UI timezone
+            try {
+                $ui_tz = new DateTimeZone($rcmail->config->get('timezone'));
+            } catch (Exception $e) {
+                $ui_tz = new DateTimeZone(DateTimeZone::UTC);
+            }
+            try {
+                $ical_tz = new DateTimeZone($ical->calendarTimeZone(true));
+            } catch (Exception $e) {
+                // Use UI timezone if the calendar have no timezone specified
+                $ical_tz = $ui_tz;
+            }
+
             // Get first event
             foreach ($ical->events() as &$event) {
-                $dtstart = $event->dtstart_array[2];
-                $dtend = $event->dtend_array[2];
-                $dtstr = $rcmail->format_date($dtstart, $combined_format) . ' - ';
-
-                // Dont double date if same
-                $df = 'Y-m-d';
-                if (date($df, $dtstart) === date($df, $dtend)) {
-                    $dtstr .= $rcmail->format_date($dtend, $time_format);
+                $is_all_day = isset($event->dtstart_array[0]['VALUE']) && $event->dtstart_array[0]['VALUE'] === 'DATE';
+
+                if ($is_all_day) {
+                    // All day events should use UI timezone to avoid turning dates
+                    $dtstart = new DateTime($event->dtstart, $ui_tz);
+                    $dtend = new DateTime($event->dtend, $ui_tz);
+                    // All day events ends at next day midnight, we should fix the day
+                    $dtend->modify('-1 day');
                 } else {
-                    $dtstr .= $rcmail->format_date($dtend, $combined_format);
+                    // Events with proper time should use calendar timezone.
+                    // Note that if DTSTART/DTEND ends with "Z", UTC is used instead of the given timezone automatically
+                    $dtstart = new DateTime($event->dtstart, $ical_tz);
+                    $dtend = new DateTime($event->dtend, $ical_tz);
```

Comment:
We should really be checking if the dates have timezone information instead ... correct me if I'm wrong, but I think it is possible that the event date and the `DTSTART`/`DTEND` may be in different timezones. In that sense, we should fall back to the event timezone only if the the date timezone is missing.

Reply:
What you can do probably is check if `dtstart_array[0]` and `dtend_array[0]` respectively have the key `TZID`. If yes, directly use `dtstart_array[2]` in `format_date`, otherwise create a new `DateTime` object using the timezone from `calendarTimeZone` if any.